Add cummax operator

anjali411 opened the pull request.

anjali411 committed:
added cummax to native_functions.yaml and ReduceOps

anjali411 committed:
added cummax to native_functions.yaml and ReduceOps

anjali411 committed:
more code

anjali411 committed:
more code

anjali411 committed:
--

anjali411 committed:
--

anjali411 committed:
merge conflict

anjali411 committed:
merge conflict

anjali411 committed:
added cummax entries in aten and THTensorMath

anjali411 committed:
added cummax entries in aten and THTensorMath

anjali411 committed:
removed non-essential code

anjali411 committed:
removed non-essential code

anjali411 committed:
--

anjali411 committed:
--

anjali411 committed:
added forward and backward fn for cummax

anjali411 committed:
added forward and backward fn for cummax

anjali411 committed:
minor updates

anjali411 committed:
minor updates

albanD commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -425,6 +425,16 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
+  if (input.dim() == 0) {
+    return input;
+  }
+  auto result = at::native::cummax(input, dim);
```
No need to recompute this, you have access to the result of the forward already by changing the `derivatives.yaml` to `self: cummax_backward(grad.to(self.scalar_type()), self, cummax_out, cummax_indices, dim)`.

anjali411 replied:
good catch! will do

albanD commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -425,6 +425,16 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
+  auto result = at::native::cummax(input, dim);
+  auto ret = std::get<0>(result);
+  auto indices = std::get<1>(result);
+  return grad * at::native::scatter(ret, dim, indices, at::ones(ret.sizes()));
```
The `scatter` here is out of place right?

anjali411 replied:
yup

albanD commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -425,6 +425,16 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
   return ret;
 }
 
+Tensor cummax_backward(const Tensor &grad, const Tensor &input, int64_t dim) {
```
If you don't define a backward for this function in `derivatives.yaml`, you need to make sure it is differentiable.
In particular, you might want to use `at::scatter` instead of the native version to make sure you can get gradients for it.

albanD commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1350,6 +1350,7 @@ class ShapePropagator {
         {
           "aten::cumprod(Tensor self, int dim, *, int? dtype) -> Tensor",
           "aten::cumsum(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::cumsum(Tensor self, int dim, *, int? dtype) -> Tensor",
```
typo? should be cummax right?

albanD commented on docs/source/tensors.rst:
```diff
@@ -219,6 +219,7 @@ view of a storage and defines numeric operations on it.
    .. automethod:: cpu
    .. automethod:: cross
    .. automethod:: cuda
+   .. automethod:: cummax
```
I guess the reason for not having an inplace version is just that we don't have one for `cumsum`?

anjali411 replied:
yeah but do you think we should have an inplace version?

albanD replied:
Maybe not if it's not needed.
It should be easy enough to add it later.

albanD reviewed (changes requested):
Looks generally good, few things to change.
High level questions:

- What is tested by `test_reduction`? Does it covers gradient/double gradients?
- How is the named tensor handling done? Does the `cummax.dimname` is automatically generated to call into `cummax`?

anjali411 committed:
minor

anjali411 committed:
minor

nairbv commented on aten/src/ATen/core/aten_interned_strings.h:
```diff
@@ -278,6 +278,7 @@ _(aten, cudnn_convolution_transpose_backward_weight) \
 _(aten, cudnn_grid_sampler) \
 _(aten, cudnn_grid_sampler_backward) \
 _(aten, cudnn_is_acceptable) \
+_(aten, cummax) \
```
@anjali411 test comment

anjali411 committed:
added test for named tensor, updated code for backwards, and other minor changes

anjali411 committed:
added test for named tensor, updated code for backwards, and other minor changes

anjali411 committed:
minor

anjali411 committed:
minor

kostmo commented:
## CircleCI build failures summary
As of commit f34d7b2:
* **3/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f34d7b27ef9531233733db044749ed0c713f1f3a).

### 3 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3966528) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75277832) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75277832))
<details>
<summary>
<code>Dec 16 20:47:10 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)'</code>
</summary>

```
Dec 16 20:47:09 gmake[2]: *** [bin/half_float_ops_test] Error 1 
Dec 16 20:47:09 gmake[1]: *** [caffe2/CMakeFiles/half_float_ops_test.dir/all] Error 2 
Dec 16 20:47:09 [ 83%] Linking CXX executable ../bin/nnpack_test 
Dec 16 20:47:09 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:47:09 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:47:09 collect2: error: ld returned 1 exit status 
Dec 16 20:47:09 gmake[2]: *** [bin/nnpack_test] Error 1 
Dec 16 20:47:09 gmake[1]: *** [caffe2/CMakeFiles/nnpack_test.dir/all] Error 2 
Dec 16 20:47:10 [ 83%] Linking CXX executable ../bin/pattern_net_transform_test 
Dec 16 20:47:10 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:47:10 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:47:10 collect2: error: ld returned 1 exit status 
Dec 16 20:47:10 gmake[2]: *** [bin/pattern_net_transform_test] Error 1 
Dec 16 20:47:10 gmake[1]: *** [caffe2/CMakeFiles/pattern_net_transform_test.dir/all] Error 2 
Dec 16 20:47:10 gmake: *** [all] Error 2 
Dec 16 20:47:10 Traceback (most recent call last): 
Dec 16 20:47:10   File "setup.py", line 755, in <module> 
Dec 16 20:47:10     build_deps() 
Dec 16 20:47:10   File "setup.py", line 316, in build_deps 
Dec 16 20:47:10     cmake=cmake) 
Dec 16 20:47:10   File "/pytorch/tools/build_pytorch_libs.py", line 62, in build_caffe2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3966520) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75278352) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75278352))
<details>
<summary>
<code>Dec 16 20:48:11 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)'</code>
</summary>

```
Dec 16 20:48:09 gmake[2]: *** [bin/bound_shape_inference_test] Error 1 
Dec 16 20:48:09 gmake[1]: *** [caffe2/CMakeFiles/bound_shape_inference_test.dir/all] Error 2 
Dec 16 20:48:09 [ 84%] Linking CXX executable ../bin/nnpack_test 
Dec 16 20:48:10 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:10 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:10 collect2: error: ld returned 1 exit status 
Dec 16 20:48:10 gmake[2]: *** [bin/nnpack_test] Error 1 
Dec 16 20:48:10 gmake[1]: *** [caffe2/CMakeFiles/nnpack_test.dir/all] Error 2 
Dec 16 20:48:10 [ 84%] Linking CXX executable ../bin/pattern_net_transform_test 
Dec 16 20:48:11 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:11 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:11 collect2: error: ld returned 1 exit status 
Dec 16 20:48:11 gmake[2]: *** [bin/pattern_net_transform_test] Error 1 
Dec 16 20:48:11 gmake[1]: *** [caffe2/CMakeFiles/pattern_net_transform_test.dir/all] Error 2 
Dec 16 20:48:11 gmake: *** [all] Error 2 
Dec 16 20:48:11 Traceback (most recent call last): 
Dec 16 20:48:11   File "setup.py", line 755, in <module> 
Dec 16 20:48:11     build_deps() 
Dec 16 20:48:11   File "setup.py", line 316, in build_deps 
Dec 16 20:48:11     cmake=cmake) 
Dec 16 20:48:11   File "/pytorch/tools/build_pytorch_libs.py", line 62, in build_caffe2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3966532) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75278493) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75278493))
<details>
<summary>
<code>Dec 16 20:48:30 FAILED: bin/conv_to_nnpack_transform_test</code>
</summary>

```
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 collect2: error: ld returned 1 exit status 
Dec 16 20:48:30 [2627/2995] Linking CXX executable bin/dead_code_elim_test 
Dec 16 20:48:30 FAILED: bin/dead_code_elim_test  
_code_elim_test.dir/opt/dead_code_elim_test.cc.o  -o bin/dead_code_elim_test  -Wl,-rpath,/pytorch/build/lib: lib/libgtest_main.a -Wl,--no-as-needed,/pytorch/build/lib/libtorch.so -Wl,--as-needed -Wl,--no-as-needed,/pytorch/build/lib/libtorch_cpu.so -Wl,--as-needed lib/libprotobuf.a lib/libc10.so /opt/intel/lib/libmkl_intel_lp64.a /opt/intel/lib/libmkl_gnu_thread.a /opt/intel/lib/libmkl_core.a -fopenmp -lpthread -lm /usr/lib/x86_64-linux-gnu/libdl.so lib/libmkldnn.a lib/libgtest.a -lpthread && : 
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 collect2: error: ld returned 1 exit status 
Dec 16 20:48:30 [2628/2995] Linking CXX executable bin/conv_to_nnpack_transform_test 
Dec 16 20:48:30 FAILED: bin/conv_to_nnpack_transform_test  
rms/conv_to_nnpack_transform_test.cc.o  -o bin/conv_to_nnpack_transform_test  -Wl,-rpath,/pytorch/build/lib: lib/libgtest_main.a -Wl,--no-as-needed,/pytorch/build/lib/libtorch.so -Wl,--as-needed -Wl,--no-as-needed,/pytorch/build/lib/libtorch_cpu.so -Wl,--as-needed lib/libprotobuf.a lib/libc10.so /opt/intel/lib/libmkl_intel_lp64.a /opt/intel/lib/libmkl_gnu_thread.a /opt/intel/lib/libmkl_core.a -fopenmp -lpthread -lm /usr/lib/x86_64-linux-gnu/libdl.so lib/libmkldnn.a lib/libgtest.a -lpthread && : 
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax(at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::native::cummax_out(at::Tensor&, at::Tensor&, at::Tensor const&, at::Dimname, c10::optional<c10::ScalarType>)' 
Dec 16 20:48:30 collect2: error: ld returned 1 exit status 
Dec 16 20:48:31 [2629/2995] Building CXX object caffe2/CMakeFiles/module_test.dir/core/module_test.cc.o 
Dec 16 20:48:31 [2630/2995] Building CXX object caffe2/CMakeFiles/transform_test.dir/core/transform_test.cc.o 
Dec 16 20:48:31 [2631/2995] Building CXX object caffe2/CMakeFiles/backend_fallback_test.dir/__/aten/src/ATen/test/backend_fallback_test.cpp.o 
Dec 16 20:48:31 [2632/2995] Building CXX object caffe2/CMakeFiles/kernel_function_legacy_test.dir/__/aten/src/ATen/core/boxing/kernel_function_legacy_test.cpp.o 
Dec 16 20:48:32 [2633/2995] Building CXX object caffe2/CMakeFiles/kernel_lambda_test.dir/__/aten/src/ATen/core/boxing/kernel_lambda_test.cpp.o 
Dec 16 20:48:32 [2634/2995] Building CXX object caffe2/CMakeFiles/net_simple_refcount_test.dir/core/net_simple_refcount_test.cc.o 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

nairbv commented on docs/source/torch.rst:
```diff
@@ -308,6 +308,7 @@ Other Operations
 .. autofunction:: cdist
 .. autofunction:: combinations
 .. autofunction:: cross
+.. autofunction:: cummax
```
I'd expect to see something in _torch_docs.py and/or _tensor_docs.py?

nairbv commented:
It's helpful to have a more detailed title/description/commit message.

Commit:
added function definitions for named tensor declarations

Commit:
added function definitions for named tensor declarations

anjali411 committed:
fixed forward and backward function logic

anjali411 committed:
fixed forward and backward function logic

anjali411 committed:
fixed dtype issue for max and scatter function in fwd and bwd fn

anjali411 committed:
fixed dtype issue for max and scatter function in fwd and bwd fn

anjali411 committed:
added autograd test, test for zero dim input

anjali411 committed:
added autograd test, test for zero dim input

anjali411 committed:
added documentation

anjali411 committed:
added documentation

anjali411 committed:
flake8 error fix

anjali411 committed:
flake8 error fix

albanD commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -425,6 +425,14 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
+    return input;
+  }
+  auto result = at::zeros(indices.sizes());
+  return result.scatter_add_(dim, indices.toType(at::kLong), grad.toType(at::kDouble));
```
The type should already be `kLong` here for indices.

albanD commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -425,6 +425,14 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
+  if (input.dim() == 0) {
+    return input;
+  }
+  auto result = at::zeros(indices.sizes());
```
Will the result dtype be correct here? It should be the same as input.

albanD commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -215,7 +215,44 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+}
 
+std::tuple<Tensor, Tensor> cummax(const Tensor& self, int64_t dim, c10::optional<ScalarType> dtype) {
+  auto out = at::empty(self.sizes(), self.options());
```
Shouldn't `out` be constructed using the given `dtype`? Otherwise, it is just ignored no?

anjali411 replied:
thanks for the catch! fixed it

albanD commented on docs/source/tensors.rst:
```diff
@@ -219,6 +219,7 @@ view of a storage and defines numeric operations on it.
    .. automethod:: cpu
    .. automethod:: cross
    .. automethod:: cuda
+   .. automethod:: cummax
```
I don't think I saw the doc string for the Tensor version. It should just point to the `torch.` version.

albanD reviewed (commented):
Looks good.
Only small updates to do !

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -215,7 +215,44 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+      toString(dtype.value()),
+      ".");
+  if(self.dim() == 0) {
+    out = self;
```
yeah I think we use NoNamesGuard so that we don't have to track names through all the complex operations and instead just update the names at the end but I guess putting that within  NoNamesGuard scope doesn't hurt either so I'll update it

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -215,7 +215,44 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+    auto out_ = self.narrow(dim, 0, 1);
+    auto indices_ = at::zeros(self.narrow(dim, 0, 1).sizes());
+    for(int64_t i=1; i<self.size(dim); i++) {
+      auto res_at_i = at::native::max(at::cat({out_.narrow(dim, i-1, 1), self.narrow(dim, i, 1)}, dim), dim);
```
I use the cat so that I can also get the indices one of the return values which the element-wise max function doesn't return I think, right?

anjali411 committed:
updated forward fn

anjali411 committed:
updated forward fn

anjali411 committed:
dtype fix code update

anjali411 committed:
dtype fix code update

anjali411 committed:
indices ret type fix -- will be kLong

anjali411 committed:
indices ret type fix -- will be kLong

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -215,7 +215,44 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+  }
+  namedinference::propagate_names(out, self);
+  namedinference::propagate_names(indices, self);
+  return std::tuple<Tensor &,Tensor &>{out, indices};
```
yeah thanks fixed it :)

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -897,6 +897,20 @@
 - func: cumprod.dimname_out(Tensor self, Dimname dim, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
   supports_named_tensor: True
 
+- func: cummax(Tensor self, int dim, *, ScalarType? dtype=None) -> (Tensor out, Tensor indices)
```
I don't think we need the dtype parameter.  It's useful in cumsum because we are potentially doing a big accumulation where you would want higher precision.  But with cummax you can always represent the answer in the dtype specified.

Also see the comment in shape_analysis below.

gchanan commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1350,6 +1350,7 @@ class ShapePropagator {
         {
           "aten::cumprod(Tensor self, int dim, *, int? dtype) -> Tensor",
           "aten::cumsum(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::cummax(Tensor self, int dim, *, int? dtype) -> (Tensor, Tensor)",
```
look at the "Requirements above", particularly: 
``` //   scalar type    : dtype if specified, preserved if floating point,
    //    otherwise long/int64	    //    otherwise long/int64```

I don't think that behavior is sensible for cummax, so it should definitely not be in this list and we should just get rid of the parameter IMO.

anjali411 replied:
oh yeah sorry didn't realize. what do you mean by the "parameter"? are you talking about the second return value(indices)?

anjali411 replied:
yeah didn't realize. thanks for pointing it out

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -216,7 +216,46 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+   {
+    NoNamesGuard guard;
+    if(self.dim() == 0) {
+      out = self;
```
you almost certainly don't want to do this (we have a longstanding action item to change the output type from `Tensor&` so folks don't make this mistake).

The semantics of "out" are that you treat the passed-in tensor as a buffer, but here you are just reassigning the reference, i..e. you are totally ignoring the buffer.  Also indices is set to self, but that's not the correct dtype is it?

anjali411 replied:
thanks for pointing it out! updated

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -216,7 +216,46 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+        auto res_at_i = at::native::max(at::cat({out_.narrow(dim, i-1, 1), self.narrow(dim, i, 1)}, dim), dim);
+        auto out_at_i = (std::get<0>(res_at_i)).unsqueeze(dim);
+        out_ = at::cat({out_, out_at_i}, dim);
+        auto max_indices_at_i = at::max(indices_.narrow(dim, i-1, 1), (i * (std::get<1>(res_at_i)).unsqueeze(dim)));
```
how come sometimes you are calling at::max and sometimes you are calling at::native::max?

anjali411 replied:
synced offline: will use at::max since there's no specific requirement for dispatch

anjali411 replied:
synced offline: will use at::max both places

anjali411 committed:
removed dtype, updated cummax fn to update out and indices in place, assigning correct dtype for out and indices

anjali411 committed:
removed dtype, updated cummax fn to update out and indices in place, assigning correct dtype for out and indices

anjali411 committed:
updated documentation

anjali411 committed:
updated documentation

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -216,7 +216,38 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+    }
+    else {
+      // update out_ and indices_ for the first values along the dimension dim
+      out.narrow(dim, 0, 1) = self.narrow(dim, 0, 1);
```
don't you need to resize the outputs first?  (Like you do in the `dim() == 0` case?).

anjali411 replied:
yeah updated!

anjali411 committed:
updated backward to ensure correct behavior for tensor with 0 elements, resize out and indices in forward and added a test for cummax

anjali411 committed:
updated backward to ensure correct behavior for tensor with 0 elements, resize out and indices in forward and added a test for cummax

anjali411 committed:
updated forward to handle cummax over a zero length dimension and added a test for backward for the same case

anjali411 committed:
updated documentation

anjali411 committed:
updated documentation

albanD commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -217,7 +217,40 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+      out.narrow(dim, 0, 1) = self.narrow(dim, 0, 1);
+      indices.narrow(dim, 0, 1).fill_(0);
+      for(int64_t i=1; i<self.size(dim); i++) {
+        auto res_at_i = at::max(at::cat({out.narrow(dim, i-1, 1), self.narrow(dim, i, 1)}, dim), dim);
```
Does at::max has the `keepDim` argument? That would avoid all the unsqueeze below.

anjali411 replied:
yeah that's a good idea. thanks, updated!

albanD reviewed (changes requested):
The test failures are relevant right?

anjali411 committed:
removed the cummax dtype reduce test and set keepdim=True for max op in cummax fwd

anjali411 committed:
removed the cummax dtype reduce test and set keepdim=True for max op in cummax fwd

albanD commented:
Also the doc for the Tensor version is missing? Or is it not needed?
Otherwise looks ready for me.
The code could be sped up but I think it's beyond this PR.

anjali411 committed:
minor

anjali411 committed:
minor

anjali411 committed:
added tensor doc

anjali411 committed:
added tensor doc

albanD commented on test/test_autograd.py:
```diff
@@ -2826,7 +2826,6 @@ def test_reduction(op, has_no_dim):
                 grad2, = torch.autograd.grad([op(x, dtype=torch.double)], [x])
                 self.assertEqual(grad1, grad2)
                 self.assertEqual(grad2.dtype, torch.float)
-
```
nit: spurious line deletion.

albanD commented on test/test_torch.py:
```diff
@@ -6259,6 +6259,7 @@ def test_scalar_check(self, device):
         # cumsum / cumprod
```
nit: "cumsum / cumprod / cummax"

albanD commented on torch/_torch_docs.py:
```diff
@@ -4150,7 +4179,8 @@ def merge_dicts(*dicts):
             [0., 4., 5., 3.],
-            [2., 1., 4., 2.]])    
+            [2., 1., 4., 2.]])
+
```
nit: spurious extra line?

albanD reviewed (changes requested):
As discussed offline, some tests for autograd are missing. Then it will be good to go.

anjali411 committed:
added tests for gradient value check, neg dim, zero dim

anjali411 committed:
minor

anjali411 committed:
minor

anjali411 committed:
minor

anjali411 committed:
minor

anjali411 committed:
added correct options for out and indices -- in response to failing cuda tests

anjali411 committed:
added correct options for out and indices -- in response to failing cuda tests

anjali411 committed:
cuda test fix

anjali411 committed:
cuda test fix

anjali411 committed:
changed cummax result name from out to values, updated test to provide unique tensor since cuda randomly choses indices for max in case of similar values, added cummax to all_operators_with_namedtuple_return list

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -217,7 +217,40 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
+std::tuple<Tensor&, Tensor&> cummax_out(Tensor& values, Tensor& indices, const Tensor& self, int64_t dim) {
+   {
+    NoNamesGuard guard;
+    values.resize_(self.sizes());
```
@albanD should I also update the options of values and indices to that of self?

anjali411 committed:
updated code to ensure out tensors and input have same dtype, device and layout. indices should be a long tensor

anjali411 committed:
updated error checks

XiaobingSuper committed:
Move rshift to Aten (#31594)

Summary:
VitalyFedyunin , this PR is about move rshift to Aten.
Benchmark script :
```
import timeit
import torch
torch.manual_seed(1)

for n, t in [(10, 100000),(1000, 10000)]:
    print('__rshift__ (a.numel() == {}) for {} times'.format(n, t))
    for device in ('cpu', 'cuda'):
        for dtype in ('torch.int8', 'torch.uint8', 'torch.int16', 'torch.int32', 'torch.int64'):
            print(f'device: {device}, dtype: {dtype}, {t} times', end='\t\t')
            print(timeit.timeit(f'a >> b\nif "{device}" == "cuda": torch.cuda.synchronize()', setup=f'import torch; a = torch.randint(0, 10, ({n},), dtype = {dtype}, device="{device}"); b = torch.randint(0, 10, ({n},), dtype = {dtype}, device="{device}")', number=t))
        for dtype in ('torch.float32', 'torch.float64'):
            print(f'device: {device}, dtype: {dtype}, {t} times', end='\t\t')
            print(timeit.timeit(f'a >> b\nif "{device}" == "cuda": torch.cuda.synchronize()', setup=f'import torch; a = torch.randn({n}, dtype = {dtype}, device="{device}"); b = torch.randn({n}, dtype = {dtype}, device="{device}")', number=t))

for n, t in [(10, 100000),(1000, 10000)]:
    print('__irshift__ (a.numel() == {}) for {} times'.format(n, t))
    for device in ('cpu', 'cuda'):
        for dtype in ('torch.int8', 'torch.uint8', 'torch.int16', 'torch.int32', 'torch.int64'):
            print(f'device: {device}, dtype: {dtype}, {t} times', end='\t\t')
            print(timeit.timeit(f'a >> b\nif "{device}" == "cuda": torch.cuda.synchronize()', setup=f'import torch; a = torch.randint(0, 10, ({n},), dtype = {dtype}, device="{device}"); b = torch.tensor(5, dtype = {dtype}, device="{device}")', number=t))
        for dtype in ('torch.float32', 'torch.float64'):
            print(f'device: {device}, dtype: {dtype}, {t} times', end='\t\t')
            print(timeit.timeit(f'a >> b\nif "{device}" == "cuda": torch.cuda.synchronize()', setup=f'import torch; a = torch.randn({n}, dtype = {dtype}, device="{device}"); b = torch.tensor(5, dtype = {dtype}, device="{device}")', number=t))
```
Device: **Tesla P100, skx-8180**
Cuda verison: **9.0.176**

Before:
```
__rshift__ (a.numel() == 10) for 100000 times
device: cpu, dtype: torch.int8, 100000 times            0.17183916084468365
device: cpu, dtype: torch.uint8, 100000 times           0.16587729007005692
device: cpu, dtype: torch.int16, 100000 times           0.16659130714833736
device: cpu, dtype: torch.int32, 100000 times           0.17177579551935196
device: cpu, dtype: torch.int64, 100000 times           0.17860156949609518
device: cpu, dtype: torch.float32, 100000 times         0.23938780091702938
device: cpu, dtype: torch.float64, 100000 times         0.22591270506381989
device: cuda, dtype: torch.int8, 100000 times           1.2709560776129365
device: cuda, dtype: torch.uint8, 100000 times          1.2692269310355186
device: cuda, dtype: torch.int16, 100000 times          1.2785452520474792
device: cuda, dtype: torch.int32, 100000 times          1.2733035255223513
device: cuda, dtype: torch.int64, 100000 times          1.2785427365452051
device: cuda, dtype: torch.float32, 100000 times                1.2980637094005942
device: cuda, dtype: torch.float64, 100000 times                1.3062487514689565
__rshift__ (a.numel() == 1000) for 10000 times
device: cpu, dtype: torch.int8, 10000 times             0.03122080024331808
device: cpu, dtype: torch.uint8, 10000 times            0.030290847644209862
device: cpu, dtype: torch.int16, 10000 times            0.024531075730919838
device: cpu, dtype: torch.int32, 10000 times            0.024743229150772095
device: cpu, dtype: torch.int64, 10000 times            0.025563121773302555
device: cpu, dtype: torch.float32, 10000 times          0.6707976600155234
device: cpu, dtype: torch.float64, 10000 times          0.5344798369333148
device: cuda, dtype: torch.int8, 10000 times            0.12768010422587395
device: cuda, dtype: torch.uint8, 10000 times           0.12681372743099928
device: cuda, dtype: torch.int16, 10000 times           0.12995595764368773
device: cuda, dtype: torch.int32, 10000 times           0.12989260721951723
device: cuda, dtype: torch.int64, 10000 times           0.12804713658988476
device: cuda, dtype: torch.float32, 10000 times         0.13013121113181114
device: cuda, dtype: torch.float64, 10000 times         0.1406280631199479
__irshift__ (a.numel() == 10) for 100000 times
device: cpu, dtype: torch.int8, 100000 times            0.3805475188419223
device: cpu, dtype: torch.uint8, 100000 times           0.36341007333248854
device: cpu, dtype: torch.int16, 100000 times           0.36908434610813856
device: cpu, dtype: torch.int32, 100000 times           0.3669992135837674
device: cpu, dtype: torch.int64, 100000 times           0.37847711704671383
device: cpu, dtype: torch.float32, 100000 times         0.4311870699748397
device: cpu, dtype: torch.float64, 100000 times         0.44503832422196865
device: cuda, dtype: torch.int8, 100000 times           1.4343859804794192
device: cuda, dtype: torch.uint8, 100000 times          1.4298221375793219
device: cuda, dtype: torch.int16, 100000 times          1.4460898758843541
device: cuda, dtype: torch.int32, 100000 times          1.4518025070428848
device: cuda, dtype: torch.int64, 100000 times          1.4456725595518947
device: cuda, dtype: torch.float32, 100000 times                1.4610810624435544
device: cuda, dtype: torch.float64, 100000 times                1.4736663019284606
__irshift__ (a.numel() == 1000) for 10000 times
device: cpu, dtype: torch.int8, 10000 times             0.05944254994392395
device: cpu, dtype: torch.uint8, 10000 times            0.058085592463612556
device: cpu, dtype: torch.int16, 10000 times            0.05094402376562357
device: cpu, dtype: torch.int32, 10000 times            0.050842881202697754
device: cpu, dtype: torch.int64, 10000 times            0.06223891582340002
device: cpu, dtype: torch.float32, 10000 times          0.7006897022947669
device: cpu, dtype: torch.float64, 10000 times          0.5614962242543697
device: cuda, dtype: torch.int8, 10000 times            0.1461706068366766
device: cuda, dtype: torch.uint8, 10000 times           0.14335164614021778
device: cuda, dtype: torch.int16, 10000 times           0.1448021186515689
device: cuda, dtype: torch.int32, 10000 times           0.14513055887073278
device: cuda, dtype: torch.int64, 10000 times           0.1439579650759697
device: cuda, dtype: torch.float32, 10000 times         0.14666561130434275
device: cuda, dtype: torch.float64, 10000 times         0.1540807681158185
```
After:
```
_rshift__ (a.numel() == 10) for 100000 times
device: cpu, dtype: torch.int8, 100000 times            0.16366520430892706
device: cpu, dtype: torch.uint8, 100000 times           0.16091545950621367
device: cpu, dtype: torch.int16, 100000 times           0.1659633992239833
device: cpu, dtype: torch.int32, 100000 times           0.1682385364547372
device: cpu, dtype: torch.int64, 100000 times           0.17289020214229822
device: cpu, dtype: torch.float32, 100000 times         0.24359441827982664
device: cpu, dtype: torch.float64, 100000 times         0.21783945057541132
device: cuda, dtype: torch.int8, 100000 times           1.2517220517620444
device: cuda, dtype: torch.uint8, 100000 times          1.260181212797761
device: cuda, dtype: torch.int16, 100000 times          1.2681935774162412
device: cuda, dtype: torch.int32, 100000 times          1.2764465296640992
device: cuda, dtype: torch.int64, 100000 times          1.294325228780508
device: cuda, dtype: torch.float32, 100000 times                1.3062216322869062
device: cuda, dtype: torch.float64, 100000 times                1.303224254399538
__rshift__ (a.numel() == 1000) for 10000 times
device: cpu, dtype: torch.int8, 10000 times             0.027045012451708317
device: cpu, dtype: torch.uint8, 10000 times            0.026978280395269394
device: cpu, dtype: torch.int16, 10000 times            0.025594274513423443
device: cpu, dtype: torch.int32, 10000 times            0.02593063935637474
device: cpu, dtype: torch.int64, 10000 times            0.02668109256774187
device: cpu, dtype: torch.float32, 10000 times          0.09746317192912102
device: cpu, dtype: torch.float64, 10000 times          0.1644029449671507
device: cuda, dtype: torch.int8, 10000 times            0.12530914042145014
device: cuda, dtype: torch.uint8, 10000 times           0.12615622486919165
device: cuda, dtype: torch.int16, 10000 times           0.12741118855774403
device: cuda, dtype: torch.int32, 10000 times           0.1284919548779726
device: cuda, dtype: torch.int64, 10000 times           0.12974756956100464
device: cuda, dtype: torch.float32, 10000 times         0.13044228963553905
device: cuda, dtype: torch.float64, 10000 times         0.13918257877230644
__irshift__ (a.numel() == 10) for 100000 times
device: cpu, dtype: torch.int8, 100000 times            0.19456563983112574
device: cpu, dtype: torch.uint8, 100000 times           0.190769555978477
device: cpu, dtype: torch.int16, 100000 times           0.2002257639542222
device: cpu, dtype: torch.int32, 100000 times           0.20456529594957829
device: cpu, dtype: torch.int64, 100000 times           0.2043834924697876
device: cpu, dtype: torch.float32, 100000 times         0.2832390898838639
device: cpu, dtype: torch.float64, 100000 times         0.2582795573398471
device: cuda, dtype: torch.int8, 100000 times           1.304957083426416
device: cuda, dtype: torch.uint8, 100000 times          1.3216373259201646
device: cuda, dtype: torch.int16, 100000 times          1.3238621400669217
device: cuda, dtype: torch.int32, 100000 times          1.333009460940957
device: cuda, dtype: torch.int64, 100000 times          1.3835567953065038
device: cuda, dtype: torch.float32, 100000 times                1.4483617274090648
device: cuda, dtype: torch.float64, 100000 times                1.4179155295714736
__irshift__ (a.numel() == 1000) for 10000 times
device: cpu, dtype: torch.int8, 10000 times             0.03196091763675213
device: cpu, dtype: torch.uint8, 10000 times            0.03048650734126568
device: cpu, dtype: torch.int16, 10000 times            0.03048624936491251
device: cpu, dtype: torch.int32, 10000 times            0.030591044574975967
device: cpu, dtype: torch.int64, 10000 times            0.031246556900441647
device: cpu, dtype: torch.float32, 10000 times          0.10918692220002413
device: cpu, dtype: torch.float64, 10000 times          0.18057993799448013
device: cuda, dtype: torch.int8, 10000 times            0.13614848721772432
device: cuda, dtype: torch.uint8, 10000 times           0.130373639985919
device: cuda, dtype: torch.int16, 10000 times           0.1332557238638401
device: cuda, dtype: torch.int32, 10000 times           0.1331850504502654
device: cuda, dtype: torch.int64, 10000 times           0.1363008264452219
device: cuda, dtype: torch.float32, 10000 times         0.1370363561436534
device: cuda, dtype: torch.float64, 10000 times         0.1442740885540843
```
Fix https://github.com/pytorch/pytorch/issues/24512 #24516  https://github.com/pytorch/pytorch/issues/24659  https://github.com/pytorch/pytorch/issues/24663
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31594

Differential Revision: D19346542

Pulled By: ezyang

fbshipit-source-id: 37dd00b86898810b850cf4769c3af8aea6d4596b

Commit:
Updating submodules

Summary:
GitHub commits:

https://github.com/hhvm/hsl/commit/b14a430062ad95c378765899f955457f3454f2f9
https://github.com/facebook/fb303/commit/c1c54260184257158d2b3feb3fc276045120c054
https://github.com/facebook/fbthrift/commit/42d18a93c4a411cef3367a35f88c48637bfa533e
https://github.com/facebook/fbzmq/commit/a4e11e872173de7824a70f977c75d5ca7fbe9f2d
https://github.com/facebook/folly/commit/25c971b0c39abc1a7a1b449e5338e1b49916da30
https://github.com/facebook/litho/commit/b2ea65322fd3b6f41541940e30034695ed41c4c0
https://github.com/facebook/mcrouter/commit/e86573b6ded555668a4b6b6215f3e702c7e4e801
https://github.com/facebook/proxygen/commit/31d721301cd410079355e4ab2a46864d57818ce5
https://github.com/facebook/rocksdb/commit/687119aeaf3799e599976330cfe673a8c4a81511
https://github.com/facebook/wangle/commit/25cad9547dfe6c2f8dbac093872643204dcb43b5
https://github.com/facebookincubator/fizz/commit/428862c04540c35881d800630ef30bcf6d7d24f3
https://github.com/facebookincubator/katran/commit/95640f80d874942439337e5d5bcffcff4559b240
https://github.com/facebookincubator/mvfst/commit/0e4db05b37545ee2f1684ffb8eb0321698659da8
https://github.com/facebookincubator/profilo/commit/5cb83de9cc2b7270eb8004361c8a61020dfad250
https://github.com/pytorch/fbgemm/commit/4fdb800074307aeb996f48746ce1341fa6db90c6

Test Plan: n/a

Reviewed By: 2d2d2d2d2d

fbshipit-source-id: bcd533c540c1170844dbf2b23538d72c95a0d304

jerryzh168 committed:
Remove qconfig_dict in top level eager mode quantization API (#31972)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31972

Since eager mode quantization requires many user modifications, we can't
consistently quantize a given model by just changing qconfig_dict, therefore
the top level `qconfig_dict` is not that useful.
fixes: https://github.com/pytorch/pytorch/issues/31549

Test Plan:
.

Imported from OSS

Differential Revision: D19330691

fbshipit-source-id: 8aee6e5249e0c14e8a363ac1a83836e88887cd7d

ljk53 committed:
skip TEST_DILL on Python2 (#32027)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32027

The test was added in #30985 for #28313. Seems the fix only works for
Python3 but doesn't work on Python2. The current Python2 CI docker image
doesn't have `dill` module installed at all so it's not captured.

I'm trying to build and push new CI docker image which has `dill` installed
and I verified it's the latest version 0.3.1.1 but the fix doesn't seem
to work and blocks me from upgrading image version. It works for Python3
docker image though...

Here is a succeeded job with old image (no dill installed):
https://app.circleci.com/jobs/github/pytorch/pytorch/4192688

Here is a failed job with new image (dill installed):
https://app.circleci.com/jobs/github/pytorch/pytorch/4192679

This PR bypasses the test for Py2 to unblock docker image change. We
can figure out a proper fix for Py2 later.

Test Plan: Imported from OSS

Differential Revision: D19341451

Pulled By: ljk53

fbshipit-source-id: d5768de8cbaf1beba8911da76f4942b8f210f2d2

ljk53 committed:
update docker image version (#31848)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31848

Trigger docker image build and bump up docker image version.

Test Plan: - Check tag at: http://ossci-docker.s3-website.us-east-1.amazonaws.com/pytorch.html

Differential Revision: D19282725

Pulled By: ljk53

fbshipit-source-id: a27b2831a92ff54d80ccbae0f18dadff0469254c

ljk53 committed:
integrate op dependency analysis process into CI

Summary:
Custom build and internal build will depend on the analysis result so
let's make sure it doesn't break.

Tested locally with LLVM-5.0, LLVM-7 and LLVM-8.

Test Plan: - check CI result

Differential Revision: D18894637

Pulled By: ljk53

fbshipit-source-id: 657854e4bed85a84907e3b6638d158823a56ec80

ljk53 committed:
gate torch_global_deps with BUILD_SHARED_LIBS flag (#32011)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32011

Run into build problem with Ninja + code analysis build as follows:
```
The install of the torch_global_deps target requires changing an RPATH from
the build tree, but this is not supported with the Ninja generator unless
on an ELF-based platform.
```

Seems we don't need build the target for static build mode?

Verified code analyzer works with the patch.

Test Plan: Imported from OSS

Differential Revision: D19336818

Pulled By: ljk53

fbshipit-source-id: 37f45a9392c45ce92c1df40d739b23954e50a13a

rohan-varma committed:
catch exceptions in ProcessGroupAgent::enqueueSend and report them. (#31023)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31023

Adds support to catch exceptions in ProcessGroupAgent::enqueueSend and
report them in the future by marking the future as completed with an exception
indicating the error. An example of when this could happen is if the receiving
side aborts when the sender is sending the message, previously, we would hang
until the timeout is hit, and the original exception would be lost.
ghstack-source-id: 96498386

Test Plan: Added a relevant unit test: `test_sender_exceptions` in rpc_test.py

Differential Revision: D18901981

fbshipit-source-id: 08de26936c4ad45b837219a247088cbea644c04c

Commit:
Scale init for batch-norm and layer-norm (#31983)

Summary:
Per discussion with Fei Tian, we need to add a `scale_init_value` to scale down the output of normalization such as batch-norm and layer-norm.

Currently we have `sparse_normalization_options` to normalize embedding pooling output. By default, scale = 1.0, we found it's better to set scale from 0.025 to 0.1 https://fb.quip.com/MiKUAibEaYhH

Besides, I am removing the tags from normalizers because it makes more sense to calculate norm ops in distributed trainers, not ps.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31983

Test Plan:
Testing LN and BN after sum-pooling --
baseline f160348514
LN: f160348609
BN: f160348710

{F226106518}

Layer norm after sum-pooling fwd_net https://fburl.com/sa4j207n
Layer norm after dot-prod fwd_net https://fburl.com/twggwyvb

## Unit Tests
Testing normalization after pooling
```
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4 -- test_sparse_pooling_batch_normalization
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4 -- test_dense_sparse_pooling_batch_normalization
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4 -- test_sparse_pooling_layer_normalization
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4 -- test_dense_sparse_pooling_layer_normalization
```

Testing normalization after dot-prod
```
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test -- test_last_layer_use_batch_norm
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test -- test_last_layer_use_layer_norm
```

Differential Revision: D19277618

Pulled By: SilunWang

fbshipit-source-id: ea323e33e3647ba55d2e808ef09d94ad7b45b934

anjali411 committed:
Updated function definition for torch.mode and torch.median in torch docs (#32003)

Summary:
Issue: https://github.com/pytorch/pytorch/issues/32002
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32003

Differential Revision: D19334306

Pulled By: anjali411

fbshipit-source-id: fe6a7cc7295b2d582a0b528f353ec64d9085e8c5

Commit:
C++ tensor indexing: add Slice / TensorIndex (#30424)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30424

`at::indexing::TensorIndex` is used for converting C++ tensor indices such as `{None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})}` into its equivalent `std::vector<TensorIndex>`, so that further tensor indexing operations can be performed using the supplied indices.

Test Plan: Imported from OSS

Differential Revision: D18695902

Pulled By: yf225

fbshipit-source-id: d73e14a411cdbec815866b02e75ffd71a9186e89

chenyangyu1988 committed:
accept url query when rank or wolrd_size is specified (#32016)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32016

The previously logic will raise exception when there is query in url when rank or world_size is specified
The fix will parse the url and stitch rank and world_size into url.query and regenerate the url.

Test Plan: f161291877

Differential Revision: D19337929

fbshipit-source-id: 6bb3a07716dda5233553804000b706052ff18db8

rohan-varma committed:
enable autograd profiler to work with RPC and RRef. (#31381)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31381

This PR adds support for being able to profile both sync and async RPCs, so that users can use the autograd profiler and be able to view metrics such as RPC latency and number of calls in the profiler output.

The way this is implemented is by using the existing `RecordFunction` class provided by the autograd profiler. We create a `RecordFunction` instance when sending an RPC, if autograd profiling is enabled. We also invoke the starting callbacks on this `RecordFunction` instance, this does things such as start the CPU timer.  This instance is then persisted across the lifetime of the RPC by attaching it to the `Future` created by the RPC. When the RPC is finished (i.e. when `future->markComplete()` is called), we run the `RecordFunction` instance's end callbacks, which among other things, stops the timer so that we get the correct RPC latency.

The `RecordFunction` and relevant callbacks in `profiler.cpp` are modified slightly to support running end callbacks from a different thread (which is needed since futures are marked as completed by a different thread than the main RPC thread). By default, the autograd profiler uses a `thread_local` list of `Events` and `thread_id`. However, since we'd like to run the `RecordFunction`'s callbacks from a different thread, we would like to access the list of `Events` created by the original thread. This is done by attaching the `thread_id` for the event to the `RecordFunction`, and then looking up the event with that thread in `all_event_lists` (see the changes in `profiler.cpp`). To ensure that the original behavior does not change in the profiler, this described behavior is only run when a user calls `setOverrideThreadId()` on the `RecordFunction` object.
ghstack-source-id: 96527291

Test Plan: Added a unit test.

Differential Revision: D19053322

fbshipit-source-id: 9a27a60c809fc4fdb16fa5d85085f3b6b21abfbb

iotamudelta committed:
Revert "Support PyTorch ROCm CI on Ubuntu18.04 (#31886)" (#31946)

Summary:
This reverts commit 4ee9c562188ae930cb2520cfce7805f55acaf968.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31946

Differential Revision: D19368391

Pulled By: bddppq

fbshipit-source-id: 63d032a5256ff4da7247fb1092be314c5b133eb6

zdevito committed:
remove list specialization from ivalue (#30734)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30734

What are specialized lists?

The IValues that hold List[int], List[Tensor], and List[AnythingElse] are different C++ types.
e.g. List[int] has a std::vector<int> while List[AnythingElse] holds a std::vector<IValue>.

Why do we have specialized lists?

When we first created the JIT we needed to bind the ATen C++ API which has std::vector<int>,
std::vector<Tensor> as inputs. The easiest way to match this API was to make our IValues contain
these same types. Conversion was just unwrapping the IValue, very easy and cheap.

What is the problem with specialized lists?

We end up with significant special cases through the compiler. Other types like Dict are not
specialized. So in the Pickler, for instance, there is a single piece of logic to handle
their serialization. For Lists, we end up with multiple cases. Furthermore, it doesn't
match Python, leading to problems along translation boundaries. Our pickle serialization
is slightly different than python, so it is harder to load objects from our IValue serialization
as Python values.

They also make it harder to provide an easy-to-use user API. We'd like to match pybind11 for C++
bindings to TorchScript. This would entail having a single torch::List class (untemplated)
that can be used to construct inputs. This is made much harder if the underlying ivalue needs
to be different depending on the type inside the list. The ideal case would be to have a constructor like

```
template<typename T>
List(std::vector<T> foo);
```

It would then set up the type tags correctly based on type T, without the need for passing tags.

Do specialized lists improve perf?

Not in a way we have been able to measure. Our major concern initially was having to translate
a std::vector<IValue> to std::vector<int> to call ATen functions. This was especially a concern
for aten::_convolution which takes a number of mostly-constant lists of integers. However,
when we measure the effect of actually having to do this conversion for an aten::_convolution,
it does not take measurable time (benchmark results below).
This is true even if you use a trivial convolution (e.g. 1x1x1), and comment out the actual convolution code.

What are the issues removing them?

This PR removes list specialization but keeps the serialization format, and IValue APIs almost exactly
the same. The only visible change is that toTensorListRef and family have turned into toTensorVector
because they now return by value a copy of the list as a vector.

Further PRs can then clean up the complexity issues that arose from speclization. This will likely
involve removing the isTensorList/isIntList functions, and refactoring the code that used them to
work generically. At some point we will also change serialization to no longer write specialized
lists in the pickle binary. This is forward incompatible, so will go in its own PR.

Benchmark:
```
import torch

import torch.nn as nn
import torch.nn.functional as F
import time

class MnistNet(nn.Module):
    def __init__(self):
        super(MnistNet, self).__init__()
        self.conv1 = nn.Conv2d(1, 1, kernel_size=1)
        self.conv2 = nn.Conv2d(1, 1, kernel_size=1)

    def forward(self, x):
        for i in range(10):
            x = F.relu(self.conv1(x))
            x = F.relu(self.conv2(x))
        return x

model = MnistNet()
x = torch.rand(1, 1, 1, 1)
r = torch.jit.trace(model, x )
r(x)
r(x)
r(x)
r(x)
print(torch.jit.last_executed_optimized_graph())

while True:
    b = time.time()
    for i in range(100):
        r(x)
    e = time.time()
    print(e - b)
```

Results (no observable difference):

```
Before (actual conv)
0.13251137733459473
0.13260436058044434
0.13276338577270508
0.1327497959136963
0.13250041007995605
0.13270330429077148
0.13290190696716309
0.13265132904052734
0.13274288177490234
0.1326758861541748
0.13253355026245117
0.13254785537719727
0.13260746002197266
0.13285017013549805
0.13264012336730957
0.132490873336792
0.13280034065246582
0.13243484497070312
0.1325232982635498
0.1326127052307129
0.13264131546020508
0.13274383544921875
0.13298296928405762
0.1326909065246582
-------------------
After (actual conv)
0.13127517700195312
0.13150334358215332
0.13092470169067383
0.13102364540100098
0.13134360313415527
0.13155555725097656
0.13314104080200195
0.13151955604553223
0.13160037994384766
0.1315293312072754
0.13137340545654297
0.13148093223571777
0.131455659866333
0.1327371597290039
0.13134026527404785
0.13152337074279785
0.13151192665100098
0.13165974617004395
0.13403725624084473
0.13251852989196777
0.13135504722595215
0.1315624713897705
0.1317615509033203
0.1314380168914795
0.13157200813293457
--------------------

The following replace the convolution operator with a no-op, to show
that even if the conv op was made faster, then we still would not see
a difference:

Before (fake conv)
0.0069539546966552734
0.0069522857666015625
0.007120847702026367
0.007344722747802734
0.007689952850341797
0.007932662963867188
0.00761723518371582
0.007501363754272461
0.007532835006713867
0.007141828536987305
0.007174253463745117
0.007114410400390625
0.007071495056152344
------------------
After (fake conv)
0.007458209991455078
0.007337093353271484
0.007268190383911133
0.007313251495361328
0.007306575775146484
0.007468700408935547
0.0073091983795166016
0.007308483123779297
0.007538318634033203
0.007356882095336914
0.007464170455932617
0.007372140884399414
```

Test Plan: Imported from OSS

Differential Revision: D18814702

Pulled By: zdevito

fbshipit-source-id: 0371c73b63068fdc12f24b801371ea90f23531a6

albanD commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -217,7 +232,41 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
 
+std::tuple<Tensor&, Tensor&> cummax_out(Tensor& values, Tensor& indices, const Tensor& self, int64_t dim) {
+  check_scalar_type_device_layout_equal(values, self);
+  check_scalar_type_device_layout_equal(indices, at::empty({0}, self.options().dtype(at::kLong)));
```
nit: do you need the `{0}` here? Can't you give `{}` so that it does not actually allocate memory?

albanD reviewed (approved):
LGTM

vishwakftw committed:
Fix typographical error in torch.triu docstring (#32067)

Summary:
below --> above

Fixes https://github.com/pytorch/pytorch/issues/32032
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32067

Differential Revision: D19355788

Pulled By: zou3519

fbshipit-source-id: dc7a2538a78cd11e72d47ad923ef50599a5a87e2

Enealor committed:
Changed clip_grad_norm_ total_norm calculation (#32020)

Summary:
Redefines the computation of the total_norm to increase performance as shown in https://github.com/pytorch/pytorch/issues/31474.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32020

Differential Revision: D19353309

Pulled By: ngimel

fbshipit-source-id: bf7530dcd39f56614a211b5f21445864d4f2e875

zasdfgbnm committed:
Display NVCC version in CI for convenience to look at

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32069

Differential Revision: D19372943

Pulled By: ezyang

fbshipit-source-id: c78e5779d4139e42df1f235db65d8c0399ffa1a2

Commit:
Install complete set of headers for ROCm build (#32076)

Summary:
This PR adds a more complete list of pytorch header files to be installed at build time. It also fixes one instance of including a header from local src directory instead of installed directory.
A more complete set of headers enable other modules to correctly work with pyTorch built for ROCm.

cc: ezyang bddppq iotamudelta
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32076

Differential Revision: D19372933

Pulled By: ezyang

fbshipit-source-id: 3b5f3241c001fa05ea448c359a706ce9a8214aa0

zasdfgbnm committed:
CUDA 8 cleanup (#32013)

Summary:
CUDA 8 is no longer supported
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32013

Differential Revision: D19372963

Pulled By: ezyang

fbshipit-source-id: e584d7d5d5908933221ea4400234b3e6e7c32e7a

nairbv committed:
Fix a typo in function name: liner -> linear

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32068

Test Plan: Imported from OSS

Differential Revision: D19373360

Pulled By: nairbv

fbshipit-source-id: 7696300b5c1dbcd7991fda3311d68807b2960982

Commit:
Fix cumprod error for tensors with zero elements (#32070)

Summary:
Currently cumprod crashes for tensors with non-empty dimensions but with zero elements, which could happen when some dimension is zero. This commit fixes the error by checking both dim() and numel() in cumprod backward
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32070

Differential Revision: D19373200

Pulled By: ezyang

fbshipit-source-id: d8ecde33f3330b40a7c611f6faa3b1d707ef2a9a

Commit:
Fix typo in config script to re-enable libtorch build and test in macOS CI (#32072)

Summary:
Currently, libtorch build and test are not running in macOS CI. This PR fixes the issue.

**Test Plan:**
Check that libtorch build and test are running again in macOS CI.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32072

Differential Revision: D19373615

Pulled By: yf225

fbshipit-source-id: 28686ef5895358a2b60db46b1946f21c58c6a18e

jerryzh168 committed:
More robust mangling (#31978)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31978

Currently we keep a `mangleIndex_` that's intenral to compilation unit and
just increment the index when we found the original name is mangled, this doesn't
guarantee the new name is not defined.
This PR fixes the problem by querying whether the new name is defined or not.
fixes: https://github.com/pytorch/pytorch/issues/31268

Test Plan:
fixes the issue

Imported from OSS

Differential Revision: D19350535

fbshipit-source-id: fe3262b2838d4208ab72e2cd4a5970b3a792ae86

zou3519 committed:
Change python/cpp docs CI to use a CPU-only image (#32102)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32102

Previously, the docs CI depended on our CUDA xenial py3 build. This
meant that the turnaround time to get signal for docs was very slow
(I've seen builds that go as much as 3 hours).

Fortunately, the docs CI do not (and should not!) rely on CUDA. This
PR changes it so that the docs CI runs on a CPU-only machine.

Fixes #29995

Test Plan:
- Check CI status on this PR by reading logs for the python and cpp docs
builds.
- I built the docs locally, once for CPU, and once for CUDA, and
verified (via diff) that the pages were exactly the same)

Differential Revision: D19374078

Pulled By: zou3519

fbshipit-source-id: 3eb36f692c3c0632d2543d3439c822d51a87b809

Commit:
Updating submodules

Summary:
GitHub commits:

https://github.com/facebook/fbthrift/commit/b5718e35c8d4cade764e8da36e699915a5ca8dce
https://github.com/facebook/folly/commit/e1af1b05501448f6a26d91d1dba3724c3aea3c0f
https://github.com/facebook/litho/commit/8a34e7f444164c8ff5523db406aeb996d9f80aa3
https://github.com/facebook/proxygen/commit/e9e70ade5bdb35a86fbb699dbf2fe7afbe2e2ab9
https://github.com/facebook/wangle/commit/d9e693ece017ba5ac3de4c5302d638bd5048f650
https://github.com/facebookincubator/fizz/commit/329347c63c1e04607d73325fa42ee752eb624f38
https://github.com/facebookincubator/katran/commit/671b5aa06432f9177a66a52c5c2247ffc3181305
https://github.com/facebookincubator/mvfst/commit/7f3bb0bf37a51aea2869af4325319e921e095158
https://github.com/facebookincubator/profilo/commit/6207e92b9be0d138c07e3bc4fde457184918fc67
https://github.com/pytorch/fbgemm/commit/d4b95d87d48cbab0e29498fcdd2b967c8e49af6c

Test Plan: n/a

Reviewed By: yns88

fbshipit-source-id: 3c9131bdee0bf8a8ca5c679a95e8ff8a6f805762

Commit:
Docs entry for the `is_quantized`

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32075

Test Plan: Imported from OSS

Differential Revision: D19353861

Pulled By: z-a-f

fbshipit-source-id: 4249216ac9a4af354a251c62181d65bc14cbfd3e

Commit:
Revert D19373615: Fix typo in config script to re-enable libtorch build and test in macOS CI

Test Plan: revert-hammer

Differential Revision:
D19373615

Original commit changeset: 28686ef58953

fbshipit-source-id: 432b04adfd9d010e1965846a386f117ebc80e013

pritamdamania committed:
Allow TCPStore to pick a port to bind to. (#31674)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31674

The motivation of this PR was to fix the problem where we would see
"Address already in use" issues for TCPStoreTest due to port conflicts. To
resolve this:

1. We can now pass in port 0 for TCPStore and retrieve the port it actually
bound to using a new getPort() API.
2. Added a `wait` flag to TCPStore constructor indicating whether or not it
should wait for workers (defaults to true).
3. Made `waitForWorkers` a public API to ensure that we can construct TCPStore
without waiting and wait for workers separately. This helps in TCPStoreTest to
ensure we can retrieve the port and pass it to the client stores.
ghstack-source-id: 96486845

Test Plan: waitforbuildbot

Differential Revision: D19240947

fbshipit-source-id: 7b1d1cb2730209fac788764845f1dbbe73d75d9b

anjali411 committed:
merge

anjali411 committed:
Revert "merge"

This reverts commit 7381dd3129784eebcc83750253b05863eafaba81, reversing
changes made to 18641061e498e84a66c93e0513780be8a4b30b3c.

anjali411 commented:
Merged by https://github.com/pytorch/pytorch/pull/32169

gmichaeljaison commented:
Is there a workaround (elegant) PyTorch implementation until we wait for this release?

gchanan commented:
@gmichaeljaison I'd recommend just using a nightly build if this is important to you.
